### PR TITLE
Perform System.exit(0) when killing the scheduler

### PR DESCRIPTION
**Pull request by mbenguig**

### CODE REVIEW

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java [hunk 1]
```diff
@@ -349,6 +351,19 @@ public boolean kill() {
 
         listener.schedulerStateUpdated(SchedulerEvent.KILLED);
 
+        // To properly exit the java scheduling process
+        new Thread(new Runnable() {
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java]: MAJOR Make this anonymous inner class a lambda [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1604)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java [hunk 2]
```diff
@@ -349,6 +351,19 @@ public boolean kill() {
...
+            @Override
+            public void run() {
+                try {
+                    Thread.sleep(SCHEDULER_EXIT_DELAY);
+                } catch (InterruptedException e) {
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java]: MAJOR Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java [hunk 3]
```diff
@@ -349,6 +351,19 @@ public boolean kill() {
...
+                    e.printStackTrace();
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java]: MINOR Use a logger to log this exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1148)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 8 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 6 minor

Watch the comments in this conversation to review them.

#### 5 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [SchedulingService.java#L622](https://github.com/ow2-proactive/scheduling/blob/bb7bd0736cb9b75f4c8e3b943025e818cc23affb/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L622): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L649](https://github.com/ow2-proactive/scheduling/blob/bb7bd0736cb9b75f4c8e3b943025e818cc23affb/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L649): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L676](https://github.com/ow2-proactive/scheduling/blob/bb7bd0736cb9b75f4c8e3b943025e818cc23affb/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L676): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L703](https://github.com/ow2-proactive/scheduling/blob/bb7bd0736cb9b75f4c8e3b943025e818cc23affb/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L703): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L728](https://github.com/ow2-proactive/scheduling/blob/bb7bd0736cb9b75f4c8e3b943025e818cc23affb/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L728): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java (hunk 1 above)
Comment (Mykhailenko) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java]: You could use lambda. Idea can help you to change it)

Comment (mbenguig) [reply scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java]: Ok thx!

File: config/scheduler/settings.ini [hunk 4]
```diff
@@ -48,6 +48,9 @@ pa.scheduler.core.nodepingfrequency=20
 # Also used by a node to retry to send the result of a task to the scheduler
 pa.scheduler.core.node.ping.attempts=1
 
+# Time in milliseconds before exiting the scheduling process
```
Comment (fviale) [inline config/scheduler/settings.ini]: when the scheduler receives a kill() request (see org.ow2.proactive.scheduler.common.Scheduler.kill)

File: scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/core/properties/PASchedulerProperties.java [hunk 5]
```diff
@@ -82,6 +82,9 @@
      * tolerated failed attempts to ping a node, before the scheduler decides to restart the task running on it */
     SCHEDULER_NODE_PING_ATTEMPTS("pa.scheduler.core.node.ping.attempts", PropertyType.INTEGER, "1"),
 
+    /** Time in milliseconds before exiting the scheduling process. */
```
Comment (fviale) [inline scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/core/properties/PASchedulerProperties.java]: same comment as above

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java (hunk 3 above)
Comment (fviale) [reply scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java]: yes, use a logger

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java [hunk 6]
```diff
@@ -349,6 +351,16 @@ public boolean kill() {
 
         listener.schedulerStateUpdated(SchedulerEvent.KILLED);
 
+        // To properly exit the java scheduling process
+        new Thread(() -> {
+            try {
+                Thread.sleep(SCHEDULER_EXIT_DELAY);
+            } catch (InterruptedException e) {
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java]: MAJOR Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

#### 5 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [SchedulingService.java#L619](https://github.com/ow2-proactive/scheduling/blob/1aa6ec105e586dd52241d07a7054adb62dff2a8a/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L619): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L646](https://github.com/ow2-proactive/scheduling/blob/1aa6ec105e586dd52241d07a7054adb62dff2a8a/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L646): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L673](https://github.com/ow2-proactive/scheduling/blob/1aa6ec105e586dd52241d07a7054adb62dff2a8a/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L673): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L700](https://github.com/ow2-proactive/scheduling/blob/1aa6ec105e586dd52241d07a7054adb62dff2a8a/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L700): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L725](https://github.com/ow2-proactive/scheduling/blob/1aa6ec105e586dd52241d07a7054adb62dff2a8a/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L725): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

#### 5 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [SchedulingService.java#L619](https://github.com/ow2-proactive/scheduling/blob/f4dd8de57f3e6bfb21c00170ecdc76cc5eef052f/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L619): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L646](https://github.com/ow2-proactive/scheduling/blob/f4dd8de57f3e6bfb21c00170ecdc76cc5eef052f/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L646): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L673](https://github.com/ow2-proactive/scheduling/blob/f4dd8de57f3e6bfb21c00170ecdc76cc5eef052f/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L673): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L700](https://github.com/ow2-proactive/scheduling/blob/f4dd8de57f3e6bfb21c00170ecdc76cc5eef052f/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L700): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L725](https://github.com/ow2-proactive/scheduling/blob/f4dd8de57f3e6bfb21c00170ecdc76cc5eef052f/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L725): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java [hunk 7]
```diff
@@ -349,6 +351,16 @@ public boolean kill() {
 
         listener.schedulerStateUpdated(SchedulerEvent.KILLED);
 
+        // To properly exit the java scheduling process
+        new Thread(() -> {
+            try {
+                Thread.sleep(SCHEDULER_KILL_DELAY);
+            } catch (InterruptedException e) {
+                logger.error("Exception occurred, fail to sleep the scheduler killing thread", e);
```
Comment (fviale) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java]: logger.warn("Scheduler.kill() was interrupted while waiting for its configured delay, now terminating the process", e)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

#### 5 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [SchedulingService.java#L620](https://github.com/ow2-proactive/scheduling/blob/3aae2cc523baa88982af681a3a1e3f54dc09c243/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L620): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L647](https://github.com/ow2-proactive/scheduling/blob/3aae2cc523baa88982af681a3a1e3f54dc09c243/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L647): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L674](https://github.com/ow2-proactive/scheduling/blob/3aae2cc523baa88982af681a3a1e3f54dc09c243/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L674): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L701](https://github.com/ow2-proactive/scheduling/blob/3aae2cc523baa88982af681a3a1e3f54dc09c243/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L701): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L726](https://github.com/ow2-proactive/scheduling/blob/3aae2cc523baa88982af681a3a1e3f54dc09c243/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L726): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

#### 5 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [SchedulingService.java#L620](https://github.com/ow2-proactive/scheduling/blob/ecdb0788ccab37375c2c9ed3f522e3d8b0bd6951/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L620): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L647](https://github.com/ow2-proactive/scheduling/blob/ecdb0788ccab37375c2c9ed3f522e3d8b0bd6951/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L647): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L674](https://github.com/ow2-proactive/scheduling/blob/ecdb0788ccab37375c2c9ed3f522e3d8b0bd6951/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L674): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L701](https://github.com/ow2-proactive/scheduling/blob/ecdb0788ccab37375c2c9ed3f522e3d8b0bd6951/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L701): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L726](https://github.com/ow2-proactive/scheduling/blob/ecdb0788ccab37375c2c9ed3f522e3d8b0bd6951/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L726): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**